google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/ClientSideStatementValueConverters.java

```diff
@@ -0,0 +1,210 @@
+              "(?is)\\A" + allowedValues.substring(1, allowedValues.length() - 1) + "\\z");
+    }
+
+    @Override
+    public Class<TimestampBound> getParameterClass() {
+      return TimestampBound.class;
+    }
+
+    @Override
+    public TimestampBound convert(String value) {
+      Matcher matcher = allowedValues.matcher(value);
+      if (matcher.find() && matcher.groupCount() >= 1) {
+        Mode mode = null;
+        int groupIndex = 0;
+        for (int group = 1; group <= matcher.groupCount(); group++) {
+          if (matcher.group(group) != null) {
+            mode = parseMode(matcher.group(group));
+            if (mode != null) {
+              groupIndex = group;
+              break;
+            }
+          }
+        }
+        switch (mode) {
+          case STRONG:
+            return TimestampBound.strong();
+          case READ_TIMESTAMP:
+            return TimestampBound.ofReadTimestamp(
+                ReadOnlyStalenessUtil.parseRfc3339(matcher.group(groupIndex + 1)));
+          case MIN_READ_TIMESTAMP:
+            return TimestampBound.ofMinReadTimestamp(
+                ReadOnlyStalenessUtil.parseRfc3339(matcher.group(groupIndex + 1)));
+          case EXACT_STALENESS:
+            try {
+              return TimestampBound.ofExactStaleness(
+                  Long.valueOf(matcher.group(groupIndex + 2)),
+                  ReadOnlyStalenessUtil.parseTimeUnit(matcher.group(groupIndex + 3)));
+            } catch (IllegalArgumentException e) {
+              throw SpannerExceptionFactory.newSpannerException(
+                  ErrorCode.INVALID_ARGUMENT, e.getMessage());
+            }
+          case MAX_STALENESS:
+            try {
+              return TimestampBound.ofMaxStaleness(
+                  Long.valueOf(matcher.group(groupIndex + 2)),
+                  ReadOnlyStalenessUtil.parseTimeUnit(matcher.group(groupIndex + 3)));
+            } catch (IllegalArgumentException e) {
+              throw SpannerExceptionFactory.newSpannerException(
+                  ErrorCode.INVALID_ARGUMENT, e.getMessage());
+            }
+          default:
+            // fall through to allow the calling method to handle this
+        }
+      }
+      return null;
+    }
+
+    private Mode parseMode(String name) {
+      for (Mode mode : Mode.values()) {
+        if (mode.name().equalsIgnoreCase(name)) {
+          return mode;
+        }
+      }
+      return null;
+    }
+  }
+
+  /** Converter for converting strings to {@link AutocommitDmlMode} values. */
+  static class AutocommitDmlModeConverter
+      implements ClientSideStatementValueConverter<AutocommitDmlMode> {
+
+    public AutocommitDmlModeConverter(String allowedValues) {}
+
+    @Override
+    public Class<AutocommitDmlMode> getParameterClass() {
+      return AutocommitDmlMode.class;
+    }
+
+    @Override
+    public AutocommitDmlMode convert(String value) {
+      for (AutocommitDmlMode mode : AutocommitDmlMode.values()) {
```

Comment:
same question for this and the below for loops

Reply:
Changed to a case-insensitive map.